File: data/readable_format/cpp/file353.cpp
#include <iostream>
#include <vector>
#include <cmath>

const double tiny = 1.0e-20;

void ludcmp(std::vector<std::vector<double>>& a, int n, std::vector<int>& indx, double& d) {
    std::vector<double> vv(n);
    d = 1.0;

    for (int i = 0; i < n; i++) {
        double aamax = 0.0;
        for (int j = 0; j < n; j++) {
            if (fabs(a[i][j]) > aamax) aamax = fabs(a[i][j]);
        }
        if (aamax == 0.0) aamax = 1.0e-6;
        vv[i] = 1.0 / aamax;
    }

    for (int j = 0; j < n; j++) {
        for (int i = 0; i < j; i++) {
            double sum = a[i][j];
            for (int k = 0; k < i; k++) {
                sum -= a[i][k] * a[k][j];
            }
            a[i][j] = sum;
        }

        double aamax = 0.0;
        int imax = 0;
        for (int i = j; i < n; i++) {
            double sum = a[i][j];
            for (int k = 0; k < j; k++) {
                sum -= a[i][k] * a[k][j];
            }
            a[i][j] = sum;

            double dum = vv[i] * fabs(sum);
            if (dum >= aamax) {
                imax = i;
                aamax = dum;
            }
        }

        if (j != imax) {
            for (int k = 0; k < n; k++) {
                double dum = a[imax][k];
                a[imax][k] = a[j][k];
                a[j][k] = dum;
            }
            d = -d;
            vv[imax] = vv[j];
        }

        indx[j] = imax;
        if (a[j][j] == 0.0) a[j][j] = tiny;
        if (j != n - 1) {
            double dum = 1.0 / a[j][j];
            for (int i = j + 1; i < n; i++) {
                a[i][j] *= dum;
            }
        }
    }
}

void printMatrix(const std::vector<std::vector<double>>& matrix) {
    for (const auto& row : matrix) {
        for (double val : row) {
            std::cout << val << " ";
        }
        std::cout << std::endl;
    }
}

int main() {
    int n = 3;
    std::vector<std::vector<double>> a = {
        {1.0, 2.0, 3.0},
        {2.0, 3.0, 4.0},
        {3.0, 4.0, 5.0}
    };
    std::vector<int> indx(n);
    double d;

    ludcmp(a, n, indx, d);

    std::cout << "Decomposed matrix:" << std::endl;
    printMatrix(a);

    return 0;
}

File: data/readable_format/fortran/file353.f90
program test_ludcmp
  implicit none
  integer, parameter :: n = 3, np = 3
  real(8) :: a(np,np), d
  integer :: indx(n)
  integer :: i, j

  ! Define a 3x3 test matrix
  data a /1.0d0, 2.0d0, 3.0d0, &
          2.0d0, 3.0d0, 4.0d0, &
          3.0d0, 4.0d0, 5.0d0/

  call ludcmp(a, n, np, indx, d)

  ! Output the decomposed matrix for verification
  do i = 1, n
    do j = 1, n
      write(*,*) 'a(', i, ',', j, ') = ', a(i,j)
    end do
  end do

contains

  subroutine ludcmp(a,n,np,indx,d)
    implicit none
    integer, intent(in) :: n, np
    real(8), dimension(np,np), intent(inout) :: a
    integer, dimension(n), intent(out) :: indx
    real(8), intent(out) :: d
    integer :: i, j, imax, k
    real(8) :: big, dum, sum, temp
    real(8), dimension(n) :: vv

    d = 1.0d0
    do i = 1, n
      big = 0.0d0
      do j = 1, n
        if (abs(a(i,j)) > big) then
          big = abs(a(i,j))
        endif
      end do
      if (big == 0.0d0) pause 'Singular matrix'
      vv(i) = 1.0d0 / big
    end do

    do j = 1, n
      do i = 1, j-1
        sum = a(i,j)
        do k = 1, i-1
          sum = sum - a(i,k)*a(k,j)
        end do
        a(i,j) = sum
      end do
      big = 0.0d0
      do i = j, n
        sum = a(i,j)
        do k = 1, j-1
          sum = sum - a(i,k)*a(k,j)
        end do
        a(i,j) = sum
        dum = vv(i) * abs(sum)
        if (dum >= big) then
          big = dum
          imax = i
        endif
      end do
      if (j /= imax) then
        do k = 1, n
          dum = a(imax,k)
          a(imax,k) = a(j,k)
          a(j,k) = dum
        end do
        d = -d
        vv(imax) = vv(j)
      endif
      indx(j) = imax
      if (a(j,j) == 0.0d0) a(j,j) = 1.0d-20
      if (j /= n) then
        dum = 1.0d0 / a(j,j)
        do i = j+1, n
          a(i,j) = a(i,j) * dum
        end do
      endif
    end do
  end subroutine ludcmp

end program test_ludcmp